### PR TITLE
[css-syntax-3][css-values-3][css-cascade-4][css-fonts-4] Formalize fetching and resource timing reporting

### LINKED ISSUE #6076: [css-fonts-4] “Font fetching requirements” doesn’t fully specify necessary requirements
The *Font fetching requirements* section of the CSS Fonts 4 spec doesn’t fully specify the requirements necessary for a UA to actually perform a font-fetch request and do something with the response.

https://drafts.csswg.org/css-fonts-4/#font-fetching-requirements

A patch (https://github.com/w3c/csswg-drafts/pull/5838) was recently merged that I wrote to add a part of the necessary requirements; but even with that change in place, there are still missing pieces:

* The spec text defines how to construct a request but should also explicitly set the request mode to `cors`.
* The spec doesn’t define what to do with the request; it needs to have something like, “Let *response* be the result of _∞fetching∞_ *request*, where _∞fetching∞_ links to the fetch algorithm in the Fetch spec.
* The spec doesn’t define what to to with the response it gets back.

@annevk

### PR DESCRIPTION
**Pull request by noamr**
Add a generic method to fetch style resources, and make use of it in
fonts & imports.

[css-syntax-3]: Make sure we have a location when parsing a stylesheet, as the location is necessary for parsing import rules.
[css-values-3]: Add a reusable method to fetch a style-originated
resource.
[css-cascade-4]: Formalize the import algorithm to use the new fetch
method.
[css-fonts-4]: Formalize font loading to use the new fetch method. Removed some of the font loading "guidelines" as they are now baked in to the algorithm.
[css-shapes-2]: Load css-shapes external images/SVG using fetch style resoure

See https://github.com/w3c/csswg-drafts/issues/562

Fixes #6076

Tests: https://github.com/web-platform-tests/wpt/pull/31344

### CODE REVIEW

**Comment by noamr** [comment]
Open-ish issues:

- The reusable method has two modes: "cors" and "no-cors". "cors" is used currently only for fonts. When "cors" is used, the referrer is set to the stylesheet's url rather than to the document's origin. I believe that this should be the general rule for anonymous cors style resources, if there are ever additional ones.

- Requires some new tests, especially for the referrer stuff, which is currently has different behaviors across browsers.

- Initial loading of the stylesheet (from `link` or `CSSOM`) is not handled in this PR, only style-originated resources.

- Not yet integrated with `[css-images]` and with SVG for `[css-shapes]`.

File: css-values-4/Overview.bs
Comment (tabatkins) [inline css-values-4/Overview.bs:998]:
```suggestion
		To <dfn export>fetch a style resource</dfn> from [=/url=] |url|, 
		given a {{CSSStyleSheet}} |sheet|, 
		a string |destination| matching a {{RequestDestination}}, 
		a "no-cors" or "cors" |corsMode|, 
		and an algorithm |processResponse| accepting a [=/response=]:
```

File: css-cascade-4/Overview.bs
Comment (tabatkins) [inline css-cascade-4/Overview.bs:280]:
```suggestion
		1. [=Fetch a style resource=] from |parsedUrl|, 
			with stylesheet |parentStylesheet|, 
			destination "style", 
			CORS mode "no-cors",
			and processResponse being the following steps given [=/response=] |response|:
```

File: css-fonts-4/Overview.bs
Comment (tabatkins) [inline css-fonts-4/Overview.bs:2951]:
```suggestion
	[=fetch a style resource|fetch=] |url|, 
	with stylesheet being |rule|'s <a spec=cssom for=CSSRule>parent CSS style sheet</a>, 
	destination "font", 
	CORS mode "cors",
	and processResponse being the following steps given [=/response=] |res|:
```

**Review by tabatkins — DISMISSED** [review]
In addition to the individual comments nits, most of which can be fixed by accepting the provided suggestions, there's some style nits where your additions don't match the surrounding spec and would need to be reformatted at some point by us.

CSSWG style is to give the algo steps their actual numbers when using Markdown numbered lists, and most importantly, we use semantic linebreaks, breaking at sentence/phrase/clause boundaries (usually after commas or before conjunctions) as feels appropriate for meaningful units of text that aren't overly long.

Otherwise looks *great*, thanks so much for filling in details that I was gradually loading into my head on my own.  Would you mind also doing the CSSOM "load a stylesheet" algo? (In this PR or separate, doesn't matter.)

**Comment by tabatkins** [comment]
> Not yet integrated with [css-images] and with SVG for [css-shapes] (any resource that I've missed?).

I *think* Colors 4, for @color-profile, is the only other place we load resources currently.

**Comment by svgeesus** [comment]
@noamr could you please link your W3C account to your GitHub account? Our IPR bot doesn't know who you are, otherwise. Thanks!

**Comment by noamr** [comment]
> @noamr could you please link your W3C account to your GitHub account? Our IPR bot doesn't know who you are, otherwise. Thanks!

I just logged in to my w3c account (noamr) and it's connected. I think I'm not in the repository's group though, there was some similar issue when I contributed to `css-shapes` a few months ago.

**Comment by noamr** [comment]
Added support for CSS shapes, still missing css-images and color profiles.

**Comment by noamr** [comment]
> > @svgeesus anything we can do about this? Not sure what's missing IPR-wise
> 
> Yes, I just sent you an email with a _non-participant commitment._ Actually I will send two, one now for the spec with the largest change and then (once you are done patching all our fetch holes) another listing all the other specs that were changed, that says "and the same for these".

So, you suggest that I finish the remaining FETCH holes as part of this PR and do an IPR thing for all these specs together?
Only `css-color` and `css-images` are now missing (CSSOM stylesheet loading will be separate).

**Comment by noamr** [comment]
> Yes, that would be perfect.

That's all done now. No additional specs planned for this PR :)

**Comment by svgeesus** [comment]
@tabatkins have your requested changes been made?

**Comment by noamr** [comment]
Ping?

**Comment by fantasai** [comment]
@svgeesus Can you rebase PRs before you merge, please? Unless it's a complicate conflicty thing that really requires a merge commit, history is a lot cleaner with a rebase. (There's a drop-down button next to the Merge button in GH that let's you choose whether to rebase or not. I'm pretty sure I've mentioned this before...)

**Comment by fantasai** [comment]
@noamr If it's not too much trouble, would you mind adding yourself to the acknowledgements section of at least some of these specs? Your call on which ones, and I'm happy to merge that PR. But I think you deserve some credit for sorting this out. :)

**Comment by noamr** [comment]
> @svgeesus Can you rebase PRs before you merge, please? Unless it's a complicated conflicty thing that really requires a merge commit, history is a lot cleaner with a rebase. (There's a drop-down button next to the Merge button in GH that let's you choose whether to rebase or not. I'm pretty sure I've mentioned this before...)

I think the Github settings allow to change the default to "Squash and merge", wouldn't that be preferable?

**Pull request merged**
